### ISSUE TITLE
How to use GPU with wasi-nn-pytorch?

### ISSUE THREAD

**Issue by 17862687921** [WasmEdge/WasmEdge#2654]
The previous examples of wasi-nn-pytorch were based on CPU, now wasi-nn-pytorch has been updated to support GPU. Could you please provide a GPU related testing example for validation and learning?

**Comment by yanghaku** [comment]
@17862687921 Hello, This is a quick example:
https://github.com/yanghaku/WasmEdge-WASINN-examples/tree/wasi-nn-0.4.0-pytorch-gpu

You can run the following commands to build and run examples  (First, you must install the latest WasmEdge and download the wasi-nn PyTorch plugin):
```shell
# clone the repo
git clone https://github.com/yanghaku/WasmEdge-WASINN-examples.git
cd WasmEdge-WASINN-examples
git checkout wasi-nn-0.4.0-pytorch-gpu
# build the example
cd pytorch-mobilenet-image/rust
cargo build --release --target=wasm32-wasi
# run the example in this folder
wasmedge --dir .:.. ./target/wasm32-wasi/release/wasmedge-wasinn-example-mobilenet-image.wasm ./mobilenet.pt ./input.jpg
```

Now the wasi-nn has updated to ```0.4.0```, And the [WasmEdge book](https://wasmedge.org/book/en/write_wasm/rust/wasinn.html) and the [wasi-nn examples](https://github.com/second-state/WasmEdge-WASINN-examples) will update these days.

**Comment by 17862687921** [comment]
> @17862687921 Hello, This is a quick example: https://github.com/yanghaku/WasmEdge-WASINN-examples/tree/wasi-nn-0.4.0-pytorch-gpu
> 
> You can run the following commands to build and run examples (First, you must install the latest WasmEdge and download the wasi-nn PyTorch plugin):
> 
> ```shell
> # clone the repo
> git clone https://github.com/yanghaku/WasmEdge-WASINN-examples.git
> cd WasmEdge-WASINN-examples
> git checkout wasi-nn-0.4.0-pytorch-gpu
> # build the example
> cd pytorch-mobilenet-image/rust
> cargo build --release --target=wasm32-wasi
> # run the example in this folder
> wasmedge --dir .:.. ./target/wasm32-wasi/release/wasmedge-wasinn-example-mobilenet-image.wasm ./mobilenet.pt ./input.jpg
> ```
> 
> Now the wasi-nn has updated to `0.4.0`, And the [WasmEdge book](https://wasmedge.org/book/en/write_wasm/rust/wasinn.html) and the [wasi-nn examples](https://github.com/second-state/WasmEdge-WASINN-examples) will update these days.

Thank you very much for your reply, but we still encounter some problems during testing. Our system is ubuntu22.04, nvidia-525.116.04, cuda11.7. The version of wasmedge installed is 0.13.1. We used the link: [https://wasmedge.org/docs/develop/build-and-run/install/#wasi-nn-plugin-with-tensorflow-lite](https://wasmedge.org/docs/develop/build-and-run/install/#wasi-nn-plugin-with-tensorflow-lite) to install the latest wasi-nn pytorch supported wasmedge , and through the link: [https://pytorch.org/get-started/locally/](https://pytorch.org/get-started/locally/) installed CUDA11.7 supporting C++/java. By executing l`dd /home/XXX/.wasmedge/plugin/libwasmedgePluginWasiNN.so`, get `libtorch.so => /home/XXX/.wasmedge/libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch /lib/libtorch.so (0x00007fc3c3c00000)`, linked. But the last step in the test process you gave: `wasmedge --dir .:.. ./target/wasm32-wasi/release/wasmedge-wasinn-example-mobilenet-image.wasm ./mobilenet.pt ./input.jpg`. After execution, an error will be reported: `wasmedge: symbol lookup error: /home/XXX/.wasmedge/lib/../plugin/libwasmedgePluginWasiNN.so: undefined symbol: _ZN5torch3jit4loadERSiN3c108optionalINS2_6DeviceEEE`. How do we need to deal with this problem?

**Comment by yanghaku** [comment]
@17862687921 Sorry, I have not noticed that in the release v0.13.1, the PyTorch plugin is a [CPU-only version with libtorch 1.8.2](https://github.com/WasmEdge/WasmEdge/blob/master/utils/wasi-nn/install-pytorch.sh).  So if you want to run it, you should build WasmEdge from the source.

```shell
git clone https://github.com/WasmEdge/WasmEdge.git
mkdir -p WasmEdge/build && pushd WasmEdge/build

## set the path to libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
export Torch_DIR=/path/to/libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
## set the path to install WasmEdge
export WASMEDGE_PATH=/path/to/install/wasmdege
## config and build
cmake .. -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_PLUGIN_WASI_NN_BACKEND=pytorch -DCMAKE_INSTALL_PREFIX=${WASMEDGE_PATH}
cmake --build . -j
cmake --install .

## check the libwasmedgePluginWasiNN.so
ldd ${WASMEDGE_PATH}/lib/wasmedge/libwasmedgePluginWasiNN.so
```

You may see the following:
```shell
$ ldd ${WASMEDGE_PATH}/lib/wasmedge/libwasmedgePluginWasiNN.so 
	linux-vdso.so.1 (0x00007ffe695bb000)
	libwasmedge.so.0 => /tmp/test/lib/libwasmedge.so.0 (0x00007f401acfe000)
	libtorch.so => /tmp/libtorch//lib/libtorch.so (0x00007f401aa00000)
	libc10.so => /tmp/libtorch//lib/libc10.so (0x00007f401a600000)
	.........
	libc10_cuda.so => /tmp/libtorch//lib/libc10_cuda.so (0x00007f4014200000)
	libtorch_cpu.so => /tmp/libtorch//lib/libtorch_cpu.so (0x00007f3ffbe00000)
	libtorch_cuda.so => /tmp/libtorch//lib/libtorch_cuda.so (0x00007f3fb6000000)
        ........
```

**Comment by 17862687921** [comment]
@yanghaku Thanks again for your help, but we still have questions to ask. Our current LLVM and LLD versions are both 17.0.0, but when we execute the above command: `cmake --build . -j`, the following error is reported:
```
[ 61%] Built target wasmedgePO
In file included from /home/XXX/program/WasmEdge/lib/aot/compiler.cpp:10:
/home/XXX/program/WasmEdge/lib/aot/llvm.h: In static member function 'static void WasmEdge::AOT::LLVM::Core::initOnce()':
/home/XXX/program/WasmEdge/lib/aot/llvm.h:91:24: error: 'LLVMGetGlobalPassRegistry' was not declared in this scope; did you mean 'LLVMGetGlobalParent'?
    91 | LLVMInitializeCore(LLVMGetGlobalPassRegistry());
       | ^~~~~~~~~~~~~~~~~~~~~~~~~
       | LLVMGetGlobalParent
/home/XXX/program/WasmEdge/lib/aot/llvm.h:91:5: error: 'LLVMInitializeCore' was not declared in this scope; did you mean 'LLVMSetInitializer'?
    91 | LLVMInitializeCore(LLVMGetGlobalPassRegistry());
       | ^~~~~~~~~~~~~~~~~~
       | LLVMSetInitializer
[ 72%] Building CXX object lib/loader/CMakeFiles/wasmedgeLoader.dir/ast/section.cpp.o
/home/XXX/program/WasmEdge/lib/aot/compiler.cpp: In function 'WasmEdge::Expect<void> {anonymous}::outputNativeLibrary(const std::filesystem::__cxx11::path&, const WasmEdge:: AOT::LLVM::MemoryBuffer&)':
/home/XXX/program/WasmEdge/lib/aot/compiler.cpp:4900:21: error: ‘lld::elf’ has not been declared
  4900 | LinkResult = lld::elf::link(
       | ^~~
gmake[2]: *** [lib/aot/CMakeFiles/wasmedgeAOT.dir/build.make:104: lib/aot/CMakeFiles/wasmedgeAOT.dir/compiler.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:577: lib/aot/CMakeFiles/wasmedgeAOT.dir/all] Error 2
gmake[1]: *** Waiting for unfinished tasks....
[ 76%] Built target wasmedgeHostModuleWasi
gmake: *** [Makefile:156:all] Error 2
```
Please, how do we need to deal with the above problem.

**Comment by hydai** [comment]
Just downgrade the LLVM to 16. I am curious that the LLVM 17 has yet to be released. Which version tag / commit hash do you use?

**Comment by 17862687921** [comment]
@hydai I am Sorry. I obtained it through the following command:
```
llvm-config --version
17.0.0git
```
```
ld.lld --version
LLD 17.0.0 (compatible with GNU linkers)
```

**Comment by hydai** [comment]
Hi @17862687921 
According to the LLVM timeline, the version 17 branch will be created on July 25th, and we will try that one after it's out.
Reference: https://discourse.llvm.org/t/llvm-17-0-0-release-planning-and-update/71762

**Comment by 17862687921** [comment]
@hydai @yanghaku Thanks again for your help, our environment is: Ubuntu22.04, LLVM 16.0.4, lld 16.0.4, CUDA Toolkit 12.1, CUDA Driver 12.1, nvidia 530.30.02, pytorch 12.1 [https://download.pytorch.org/libtorch/nightly/cu121/libtorch-shared-with-deps-latest.zip](https://download.pytorch.org/libtorch/nightly/cu121/libtorch-shared-with-deps-latest.zip). But when we execute the above case `WasmEdge-WASINN-examples`, the statement `wasmedge --dir =/home/XXX/program/install/wasmedge/lib/wasmedge ./target/wasm32-wasi/release/wasmedge-wasinn-example-mobilenet-image.wasm ./mobilenet.pt ./input.jpg`
After that, report an error: 
```thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Custom { kind: Uncategorized, error: "failed to find a pre-opened file descriptor through which \"./mobilenet.pt\" could be opened" })', src/main.rs:18:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-07-11 04:35:54.705] [error] execution failed: unreachable, Code: 0x89
[2023-07-11 04:35:54.705] [error] In instruction: unreachable (0x00) , Bytecode offset: 0x000a62c3
[2023-07-11 04:35:54.705] [error] When executing function name: "_start```
Please let me know how we need to solve the above problem, thank you again for your help.

**Comment by yanghaku** [comment]
The program cannot find `mobilenet.pt`. You can see how to use the ```--dir``` option to set directory mapping in WasmEdge.

```shell
$ wasmedge --help | grep dir -A2
--dir
		Binding directories into WASI virtual filesystem. Each directory can be
		specified as --dir `host_path`. You can also map a guest directory to a host
		directory by --dir `guest_path:host_path`, where `guest_path` specifies the
		path that will correspond to `host_path` for calls like `fopen` in the
		guest.The default permission is `readwrite`, however, you can use --dir
		`guest_path:host_path:readonly` to make the mapping directory become a read
		only mode.
```

**Comment by 17862687921** [comment]
@hydai @yanghaku Thank you very much for your help, we are using 
```
git clone https://github.com/WasmEdge/WasmEdge.git
mkdir -p WasmEdge/build && pushd WasmEdge/build

## set the path to libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
export Torch_DIR=/path/to/libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
## set the path to install WasmEdge
export WASMEDGE_PATH=/path/to/install/wasmdege
##config and build
cmake .. -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_PLUGIN_WASI_NN_BACKEND=pytorch -DCMAKE_INSTALL_PREFIX=${WASMEDGE_PATH}
cmake --build . -j
cmake --install .
```
After building wasmedge, but after executing the command 
`ldd {WASMEDGE_PATH}/lib/wasmedge/libwasmedgePluginWasiNN.so`
 the following problems were found:
```
...
libwasmedge.so.0 => not found
...
```
Please tell us how we need to deal with this problem, thank you again for your help.

**Comment by yanghaku** [comment]
```shell
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${WASMEDGE_PATH}/lib/
```

**Comment by 17862687921** [comment]
@hydai @yanghaku Sorry, we encountered a problem while using wasi-nn-pytorch. When we convert the trained model (semantic segmentation, object detection) into a pt file through the following code ([https://github.com/yanghaku/WasmEdge-WASINN-examples/blob/wasi-nn-0.4.0-pytorch-gpu/pytorch-mobilenet-image/gen_mobilenet_model.py](https://github.com/yanghaku/WasmEdge-WASINN-examples/blob/wasi-nn-0.4.0-pytorch-gpu/pytorch-mobilenet-image/gen_mobilenet_model.py)),

```
#pyton
sm = torch.jit.script(model)
if not os.path.exists("mobilenet.pt"):
        sm.save("mobilenet.pt")
load_sm = jit.load("mobilenet.pt")
```

the original training result was a list type (model output result), but it was converted to a torch script type (i.e. mobilenet.pt), and the output result was a tuple type.

However, the system prompts `[WASI-NN] PyTorch backend only supports output a tensor or a list of tensors`. How can we solve this problem by using wasi-nn-pytorch to achieve semantic segmentation and object detection testing. Thanks.

**Comment by yanghaku** [comment]
@17862687921 I think you can check the model output type.  In WasmEdge source code:

```c++
if (RawOutput.isTensorList()) {
  // ... save outputs.
} else if (RawOutput.isTuple()) {
  // ... save outputs.
} else if (RawOutput.isTensor()) {
  // ... save outputs.
} else {
  spdlog::error("[WASI-NN] PyTorch backend only supports output a tensor or a list of tensor");
  return ErrNo::InvalidArgument;
}
```

**Comment by 17862687921** [comment]
@yanghaku I can use wasi-nn in normal state, but this error is reported in sudo state
```
$ wasmedge --dir .:. semantic.so semantic.pt semantic.jpg
successful

$ sudo wasmedge --dir .:. semantic.so semantic.pt semantic.jpg
[error] WASI-NN plugin not installed. Please install the plugin and restart WasmEdge.
Error: BackendError(InvalidArgument)
```
But simple commands can be executed
```
$ wasmedge -v
wasmedge version 0.13.3-20-g175fbcad
$ sudo wasmedge -v
wasmedge version 0.13.3-20-g175fbcad
```
I built wasmedge in the following way
```
git clone https://github.com/WasmEdge/WasmEdge.git
mkdir -p WasmEdge/build && pushd WasmEdge/build

## set the path to libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
export Torch_DIR=/path/to/libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
## set the path to install WasmEdge
export WASMEDGE_PATH=/path/to/install/wasmdege
##config and build
cmake .. -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_PLUGIN_WASI_NN_BACKEND=pytorch -DCMAKE_INSTALL_PREFIX=${WASMEDGE_PATH}
cmake --build . -j
cmake --install .
```
How can I solve this problem? Thank you again for your help.

**Comment by yanghaku** [comment]
You need to set the environment variable `WASMEDGE_PLUGIN_PATH`.

**Comment by wuqingshan2010** [comment]
$ wasmedge 
[error]: terminate called after throwing an instance of 'std::out_of_range' what():  basic_string_view::substr: __pos (which is 18446744073709551615) > __size (which is 55)
Aborted (core dumped)
I built wasmedge in the following way
git clone https://github.com/WasmEdge/WasmEdge.git
mkdir -p WasmEdge/build && pushd WasmEdge/build
## set the path to libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
export Torch_DIR=/path/to/libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117/libtorch
## set the path to install WasmEdge
export WASMEDGE_PATH=/path/to/install/wasmdege
##config and build
cmake .. -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_PLUGIN_WASI_NN_BACKEND=pytorch -DCMAKE_INSTALL_PREFIX=${WASMEDGE_PATH}
cmake --build . -j
cmake --install .

**Comment by hydai** [comment]
Hi @wuqingshan2010 
Could you please provide the environmental information?

**Comment by wuqingshan2010** [comment]
> Hi @wuqingshan2010 Could you please provide the environmental information?

ubuntu 20.04
cuda toolkit 11.7
libtorch-cxx11-abi-shared-with-deps-2.0.1+cu117
apt install llvm-12-dev liblld-12-dev

**Comment by 17862687921** [comment]
Does wasmedge currently support batch image classification tasks? For example, if it is possible to complete the classification task of 100 images (batch_size=100) at the same time? Can you give usage examples? Thanks.

**Comment by juntao** [comment]
Do you mean to process images one by one without reloading the model? Something like this:

https://github.com/WasmEdge/mediapipe-rs#use-the-session-to-speed-up

You can look into the mediapipe-rs project above to see how it is implemented.

**Comment by 17862687921** [comment]
Sorry, no. Like `cifar_train = torch.utils.data.DataLoader(cifar_train, batch_size=128, shuffle=True)` in python, I hope to process a large amount of image data at once, convert it into a tensor, and then perform subsequent model inference.

**Comment by juntao** [comment]
The `batch_size` specifies how much data it sends to GPU to process. The WASI NN API is too high level for this type of details (in fact, WASI NN does not even if it is a GPU or CPU).

I think it is possible support this type of configuration options on the PyTorch backend through an additional function in WASI NN (ie to set configuration and then inference). But it is not done yet. You are welcome to contribute if you have bandwidth. Thanks.

**Comment by hydai** [comment]
Looks like the issue is resolved. Closing it. Please feel free to re-open it if the issue still exists.
